pilot/pkg/security/authz/builder/extauthz.go

```diff
@@ -0,0 +1,318 @@
+		}
+		var parsed *builtExtAuthz
+		var err error
+		// TODO(yangminzhu): Refactor and cache the ext_authz config.
+		switch p := config.Provider.(type) {
+		case *meshconfig.MeshConfig_ExtensionProvider_EnvoyExtAuthzHttp:
+			parsed, err = buildExtAuthzHTTP(p.EnvoyExtAuthzHttp)
+		case *meshconfig.MeshConfig_ExtensionProvider_EnvoyExtAuthzGrpc:
+			parsed, err = buildExtAuthzGRPC(p.EnvoyExtAuthzGrpc)
+		default:
+			err = fmt.Errorf("unsupported extension provider: %s", config.Name)
+		}
+		if err != nil {
+			errs = multierror.Append(errs, fmt.Errorf("failed to parse extension provider %s: %v", config.Name, err))
+		}
+		resolved[config.Name] = parsed
+	}
+	ret, found := resolved[provider]
+	if !found {
+		var li []string
+		for p := range resolved {
+			li = append(li, p)
+		}
+		errs = multierror.Append(fmt.Errorf("extension provider %s not found, available providers are %v", provider, li))
+	}
+	if errs != nil {
+		return nil, errs
+	}
+
+	authzLog.Debugf("Resolved provider %s to config: %v", provider, spew.Sdump(ret))
+	return ret, nil
+}
+
+func buildExtAuthzHTTP(config *meshconfig.MeshConfig_ExtensionProvider_EnvoyExternalAuthorizationHttpProvider) (*builtExtAuthz, error) {
+	var errs error
+	port, err := parsePort(config.Port)
+	if err != nil {
+		errs = multierror.Append(errs, err)
+	}
+	hostname, cluster, err := parseService(config.Service, port)
+	if err != nil {
+		errs = multierror.Append(errs, err)
+	}
+	status, err := parseStatusOnError(config.StatusOnError)
+	if err != nil {
+		errs = multierror.Append(errs, err)
+	}
+	if config.PathPrefix != "" {
+		if _, err := url.Parse(config.PathPrefix); err != nil {
+			errs = multierror.Append(errs, fmt.Errorf("invalid pathPrefix %q: %v", config.PathPrefix, err))
+		}
+		if !strings.HasPrefix(config.PathPrefix, "/") {
+			errs = multierror.Append(errs, fmt.Errorf("pathPrefix must begin with `/`, found: %q", config.PathPrefix))
+		}
+	}
+	if errs != nil {
+		return nil, errs
+	}
+
+	return generateHTTPConfig(hostname, cluster, status, config), nil
+}
+
+func buildExtAuthzGRPC(config *meshconfig.MeshConfig_ExtensionProvider_EnvoyExternalAuthorizationGrpcProvider) (*builtExtAuthz, error) {
+	var errs error
+	port, err := parsePort(config.Port)
+	if err != nil {
+		errs = multierror.Append(errs, err)
+	}
+	_, cluster, err := parseService(config.Service, port)
+	if err != nil {
+		errs = multierror.Append(errs, err)
+	}
+	status, err := parseStatusOnError(config.StatusOnError)
+	if err != nil {
+		errs = multierror.Append(errs, err)
+	}
+	if errs != nil {
+		return nil, errs
+	}
+
+	return generateGRPCConfig(cluster, config.FailOpen, status), nil
+}
+
+func parseService(service string, port int) (hostname string, cluster string, err error) {
+	if service == "" {
+		return "", "", fmt.Errorf("service must not be empty")
+	}
+	parts := strings.Split(service, "/")
+	if len(parts) > 2 {
+		return "", "", fmt.Errorf("service not in format [<namespace>/]<service>, found: %s", service)
```

Comment:
Why is this namespace/service and not namespace/FQDN? doesn't that break ability to use ServiceEntry?

Reply:
good question, I think we're trying to follow with the more general convention to refer to other (k8s or Istio) resource by using the name directly, and the indirection gives more flexibility (e.g. the user doesn't need to specify the cluster.local part)

This works well with k8s service but not with ServiceEntry because you can have multiple hosts. I think we will probably support only 1 host in ServiceEntry if it is being used in extension provider here. cc @louiscryan